### PR TITLE
Fixed bugs when trying to use very small datasets

### PR DESCRIPTION
**Pull request by snarayan21**
## Description of changes:

Previously, our sample partitioning algorithm could not handle datasets with a very small number of samples.
* now it can
* added a warning
Using small datasets also made per-stream batching throw weird errors, further also causing py1e to experience issues.
* a small bug in per stream batching has been addressed
* added a warning

Added unit tests for the partition algorithm to ensure that it works as expected on very small numbers of samples.

This fixes the issue lol:
Screenshot 2023-12-04 at 3 01 28 PM

## Issue #, if available:
STR-138

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [x] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [x] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#submitting-a-contribution))
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

File: streaming/base/batching/per_stream.py [hunk 1]
```diff
@@ -94,7 +94,13 @@ def generate_work_per_stream_batching(dataset: StreamingDataset, world: World, e
             logger.warning(
                 'Because of the `per_stream` batching method, some batches with an inadequate \
                 number of samples from stream with index ' + str(stream_idx) + ' will be dropped.')
-        batches_from_partitions.append(global_batches_inorder[:num_full_batches])
+        if num_full_batches > 0:
+            batches_from_partitions.append(global_batches_inorder[:num_full_batches])
+        else:
+            logger.warning('Stream with index ' + str(stream_idx) +
```
Comment (karan6181) [inline streaming/base/batching/per_stream.py:100]: Use f-string and `+` to concatenate long strings like [this](https://github.com/mosaicml/streaming/blob/main/streaming/base/dataset.py#L388)?

File: streaming/base/partition/orig.py [hunk 2]
```diff
@@ -80,37 +81,53 @@ def get_partitions_orig(num_samples: int,
...
+        ids[:, samples_per_canonical_node - 1:samples_per_canonical_node] -= are_rows_short
+
+        # If padding we needed, repeat samples to populate it.
+        if padding:
+            ids[:, -padding:] = ids[:,
+                                    -padding - node_ratio + 1 - padding:-padding - node_ratio + 1]
+    else:
+        # Num samples is less than the number of canonical nodes. The dataset is very small, and
+        # samples will be repeated extensively.
+        warnings.warn(f'Attempting to partition {num_samples} samples over {num_canonical_nodes} \
```
Comment (karan6181) [inline streaming/base/partition/orig.py:120]: Is there any recommendation we need to provide here? Such as lowering NCN or something else?

File: streaming/base/partition/orig.py [hunk 3]
```diff
@@ -126,10 +143,23 @@ def get_partitions_orig(num_samples: int,
...
+    more_samples_than_ranks = ids.shape[1] > ranks_per_node
     if overflow:
         underflow = ranks_per_node - overflow
-        last = ids[:, -ranks_per_node - underflow + 1:-ranks_per_node + 1]
+        if more_samples_than_ranks:
+            last = ids[:, -ranks_per_node - underflow + 1:-ranks_per_node + 1]
+        else:
+            # There are less samples than ranks. Usually, we pad by trying to ensure that the same
+            # samples don't get repeated over and over, but with in this case, we are forced to.
+            warnings.warn(f'Attempting to partition {ids.shape[1]} samples over {ranks_per_node} \
```
Comment (karan6181) [inline streaming/base/partition/orig.py:154]: Same comment here. Any recommendations we can provide?

File: tests/test_partition.py [hunk 4]
```diff
@@ -34,6 +34,22 @@ def test_partition_walk(partition_algo: str):
         assert x.shape == (22, 8, 8, 1, 10)
 
 
+@pytest.mark.parametrize('num_samples', [1, 4])
+@pytest.mark.parametrize('num_canonical_nodes', [1, 4])
+@pytest.mark.parametrize('num_physical_nodes', [1, 4])
+@pytest.mark.parametrize('ranks_per_node', [1, 8])
+@pytest.mark.parametrize('workers_per_rank', [1, 8])
+@pytest.mark.parametrize('batch_size', [4])
+def test_partition_small_num_samples(num_samples: int, num_canonical_nodes: int,
```
Comment (karan6181) [inline tests/test_partition.py:43]: Is it possible to test the warning messages using `with pytest.warns`?

File: streaming/base/partition/orig.py (hunk 2 above)
Comment (snarayan21) [reply streaming/base/partition/orig.py:120]: yeah recommendation here is to lower NCN, and check if the dataset has the expected number of samples. added in.

File: streaming/base/partition/orig.py (hunk 3 above)
Comment (snarayan21) [reply streaming/base/partition/orig.py:154]: In this case, there's not much to do except check if your dataset does have the number of samples you expected. added that in.

**Review by karan6181 — APPROVED** [review]

**Pull request merged**